topic-operator/src/main/java/io/strimzi/operator/topic/KafkaStreamsTopicStore.java

```diff
@@ -0,0 +1,114 @@
+import io.vertx.core.Future;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+
+/**
+ * TopicStore based on Kafka Streams and
+ * Apicurio Registry's gRPC based Kafka Streams ReadOnlyKeyValueStore
+ */
+public class KafkaStreamsTopicStore implements TopicStore {
+    private static final Logger log = LoggerFactory.getLogger(KafkaStreamsTopicStore.class);
+
+    private final ReadOnlyKeyValueStore<String, Topic> topicStore;
+
+    private final String storeTopic;
+    private final ProducerActions<String, TopicCommand> producer;
+
+    private final BiFunction<String, String, CompletionStage<Integer>> resultService;
+
+    public KafkaStreamsTopicStore(
+            ReadOnlyKeyValueStore<String, Topic> topicStore,
+            String storeTopic,
+            ProducerActions<String, TopicCommand> producer,
+            BiFunction<String, String, CompletionStage<Integer>> resultService) {
+        this.topicStore = topicStore;
+        this.storeTopic = storeTopic;
+        this.producer = producer;
+        this.resultService = resultService;
+    }
+
+    public static Throwable toThrowable(Integer index) {
+        if (index == null) {
+            return null;
+        }
+        switch (index) {
+            case 1:
+                return new TopicStore.EntityExistsException();
+            case 2:
+                return new TopicStore.NoSuchEntityExistsException();
+            case 3:
+                return new TopicStore.InvalidStateException();
+            default:
+                throw new IllegalStateException("Invalid index: " + index);
+        }
+    }
+
+    public static Integer toIndex(Class<? extends Throwable> ct) {
+        if (ct.equals(TopicStore.EntityExistsException.class)) {
+            return 1;
+        }
+        if (ct.equals(TopicStore.NoSuchEntityExistsException.class)) {
+            return 2;
+        }
+        if (ct.equals(TopicStore.InvalidStateException.class)) {
+            return 3;
+        }
+        throw new IllegalStateException("Unexpected value: " + ct);
+    }
+
+    @Override
+    public Future<Topic> read(TopicName name) {
+        try {
+            Topic topic = topicStore.get(name.toString());
+            return Future.succeededFuture(topic);
+        } catch (Throwable t) {
+            return Future.failedFuture(t);
+        }
+    }
+
+    private Future<Void> handleTopicCommand(TopicCommand cmd) {
+        String key = cmd.getKey();
+        CompletionStage<Throwable> result = resultService.apply(key, cmd.getUuid())
+                .thenApply(KafkaStreamsTopicStore::toThrowable);
+        // Kafka Streams can re-balance in-between these two calls ...
+        producer.apply(new ProducerRecord<>(storeTopic, key, cmd))
```

Comment:
genuine question because I don't know Apicurio ... what's the value of using this "apicurio" producer instead of a plain Kafka producer? Aren't we going just to send a message in the store topic and then the Kafka Streams application takes care of it storing information in the topic store?

Reply:
It's just a bit more resilient then plain Kafka producer -- see the code.
It also transparently handles fatal errors in callback as well.
And it uses CS as a callback API, which makes it easier to use -- when you have a CS chain.
(here we actually don't take full advantage of this)

Reply:
My concern is that we are adding an apicurio dependency in the topic operator just for this. Or apicurio stuff is used even somewhere else?

Reply:
I'll let @tombentley work out the rf value algorithm

Reply:
Yes, in a few classes.
Mostly in order to easily abstract away local vs distributed storage + async function.
And some utils - concurrency and serde.